### DISCUSSION TITLE
why is double colon used instead of single colon

### DISCUSSION THREAD

**Discussion by LeoQuote** [prometheus-operator/kube-prometheus#1174] [Q&A]
In this part of code, https://github.com/prometheus-operator/kube-prometheus/blob/4d8104817d6c6bb3f095bb2a17ad86737d2d2e33/jsonnet/kube-prometheus/main.libsonnet#L16 double colon was used, but I cannot get any info about double colon in jsonnet except for this archived project https://github.com/ksonnet/ksonnet-lib/blob/master/docs/jsonnetIntro.md#functions is that obsolete or just typo?

**Comment by paulfantom** [comment]
double colon hides object during manifestation.

From https://jsonnet.org/learning/tutorial.html, section "Object-Orientation":

> Hidden fields, defined with ::, which do not appear in generated JSON

**Reply by LeoQuote** [reply]
Thanks for your explaining, I wrote a config file for kube-prom like this. 
```
    values+:: {
      common+: {
        namespace: 'monitoring',
      },
    },
```
but later I found it can also work when using single colon, like :

```
    values+: {
      common+: {
        namespace: 'monitoring',
      },
    },
```

how did that happen? since it can also work with single colon, why would I use double colon?

**Reply by ArthurSens** [reply]
If you are using the examples in this repository, they explicit skip the `values` object when printing:
```
local kp =
  (import 'kube-prometheus/main.libsonnet') +
  {
    values+:: {
      common+: {
        namespace: 'monitoring',
      },
    },
  };

{ 'setup/0namespace-namespace': kp.kubePrometheus.namespace } +
{
  ['setup/prometheus-operator-' + name]: kp.prometheusOperator[name]
  for name in std.filter((function(name) name != 'serviceMonitor' && name != 'prometheusRule'), std.objectFields(kp.prometheusOperator))
} +
// serviceMonitor and prometheusRule are separated so that they can be created after the CRDs are ready
{ 'prometheus-operator-serviceMonitor': kp.prometheusOperator.serviceMonitor } +
{ 'prometheus-operator-prometheusRule': kp.prometheusOperator.prometheusRule } +
{ 'kube-prometheus-prometheusRule': kp.kubePrometheus.prometheusRule } +
{ ['alertmanager-' + name]: kp.alertmanager[name] for name in std.objectFields(kp.alertmanager) } +
{ ['blackbox-exporter-' + name]: kp.blackboxExporter[name] for name in std.objectFields(kp.blackboxExporter) } +
{ ['grafana-' + name]: kp.grafana[name] for name in std.objectFields(kp.grafana) } +
{ ['kube-state-metrics-' + name]: kp.kubeStateMetrics[name] for name in std.objectFields(kp.kubeStateMetrics) } +
{ ['kubernetes-' + name]: kp.kubernetesControlPlane[name] for name in std.objectFields(kp.kubernetesControlPlane) }
{ ['node-exporter-' + name]: kp.nodeExporter[name] for name in std.objectFields(kp.nodeExporter) } +
{ ['prometheus-' + name]: kp.prometheus[name] for name in std.objectFields(kp.prometheus) } +
{ ['prometheus-adapter-' + name]: kp.prometheusAdapter[name] for name in std.objectFields(kp.prometheusAdapter) }
```

Notice that it directly calls the child objects, e.g. kp.prometheus, kp.alertmanager, kp.nodeExporter, etc. Try this:
```
local kp =
  (import 'kube-prometheus/main.libsonnet') +
  {
    values+:: {
      common+: {
        namespace: 'monitoring',
      },
    },
   prometheus+:: {}
  };

{ ['prometheus-' + name]: kp.prometheus[name] for name in std.objectFields(kp.prometheus) } 
```
and 
```
local kp =
  (import 'kube-prometheus/main.libsonnet') +
  {
    values+:: {
      common+: {
        namespace: 'monitoring',
      },
    },
   prometheus+: {}
  };

{ ['prometheus-' + name]: kp.prometheus[name] for name in std.objectFields(kp.prometheus) } 
```

The first one should not create any manifests because of the double double collon `::`, while the other will generate because there is only one

**Reply by LeoQuote** [reply]
I'm still puzzled, so double colon `::` hides some field and prevent this resource being created, why would I write it in the first place ? why dont I just delete it?

**Reply by ArthurSens** [reply]
Because it holds information used on other places of kube-prometheus. See: https://github.com/prometheus-operator/kube-prometheus/blob/main/jsonnet/kube-prometheus/main.libsonnet#L37-L134

On your jsonnet file that is consuming `main.libsonnet`, you can override some fields in the "values" object, and even though it is not turned into its own file, it will configure all the others

**Reply by LeoQuote** [reply]
Sorry, can I just ask a specific question, why is the example in https://github.com/prometheus-operator/kube-prometheus/blob/main/README.md#defining-the-servicemonitor-for-each-additional-namespace is using 
```
    prometheus+:: {
      namespaces+: ['my-namespace', 'my-second-namespace'],
    },
```
while the example here https://github.com/prometheus-operator/kube-prometheus/blob/main/README.md#adding-additional-namespaces-to-monitor is using 

```
    prometheus+: {
      namespaces+: ['my-namespace', 'my-second-namespace'],
    },
```

Are these two config all configring https://github.com/prometheus-operator/kube-prometheus/blob/4d8104817d6c6bb3f095bb2a17ad86737d2d2e33/jsonnet/kube-prometheus/components/prometheus.libsonnet#L12 this fields? if these two style works exactly the same, can I just use `+:` for all fields?

**Reply by LeoQuote** [reply]
https://github.com/prometheus-operator/kube-prometheus/discussions/1174#discussioncomment-823616 I tried the two examples you provided, these two example produce exactly the same manifests. 

```
/kube-prom $ ls manifests/
prometheus-clusterRoleBinding.yaml   prometheus-roleBindingConfig.yaml              prometheus-serviceMonitor.yaml
prometheus-clusterRole.yaml          prometheus-roleBindingSpecificNamespaces.yaml  prometheus-service.yaml
prometheus-podDisruptionBudget.yaml  prometheus-roleConfig.yaml                     setup
prometheus-prometheusRule.yaml       prometheus-roleSpecificNamespaces.yaml
prometheus-prometheus.yaml           prometheus-serviceAccount.yaml
```
